main.cpp

```diff
@@ -288,436 +236,179 @@ struct HeapA
+        *value += rhs;
+        return *this;
     }
-    
-    return *this;
-}
 
-FloatType& FloatType::apply(void (*f)(float&))
-{
-    if (f)
+    Numeric<Type>& operator-=(Type rhs)
     {
-        f(*value);
+        *value -= rhs;
+        return *this;
     }
-    
-    return *this;
-}
-// ======================================== Double primitive ==========
-DoubleType& DoubleType::add(double rhs)
-{
-    *value += rhs;
-    return *this;
-}
-
-DoubleType& DoubleType::subtract(double rhs)
-{
-    *value -= rhs;
-    return *this;
-}
 
-DoubleType& DoubleType::multiply(double rhs)
-{
-    *value *= rhs;
-    return *this;
-}
-
-DoubleType& DoubleType::divide(double rhs)
-{
-    if (rhs == 0.0)
-        std::cout << "warning: floating point division by zero!" << std::endl;
-
-    *value /= rhs;
-    return *this;        
-}
-// ======================================== Double operators ==========
-DoubleType::operator double() const { return *value; }
-
-DoubleType& DoubleType::operator+=(double rhs)
-{
-    *value += rhs;
-    return *this;
-}
-
-DoubleType& DoubleType::operator-=(double rhs)
-{
-    *value -= rhs;
-    return *this;
-}
-
-DoubleType& DoubleType::operator*=(double rhs)
-{
-    *value *= rhs;
-    return *this;
-}
-
-DoubleType& DoubleType::operator/=(double rhs)
-{
-    if (rhs == 0.0)
-        std::cout << "warning: floating point division by zero!" << std::endl;
-
-    *value /= rhs;
-    return *this;        
-}
-// ======================================== Double Apply func =========
-DoubleType& DoubleType::apply(std::function<DoubleType&(double&)> f)
-{
-    if (f)
+    Numeric<Type>& operator*=(Type rhs)
     {
-        return f(*value);
+        *value *= rhs;
+        return *this;
     }
-    
-    return *this;
-}
 
-DoubleType& DoubleType::apply(void (*f)(double&))
-{
-    if (f)
+    template <typename ArgumentType>
+    Numeric<Type>& operator/=(ArgumentType rhs)
     {
-        f(*value);
+        if constexpr (std::is_same<Type, int>::value)
+        {
+            if constexpr (std::is_same<ArgumentType, int>::value)
+            {
+                if (rhs == 0)
+                {
+                    std::cout << "error: integer division by zero is an error and will crash the program!" << std::endl;
+                    return *this;
+                }
+            }
+
+            else if (std::numeric_limits<ArgumentType>::epsilon() > std::abs(rhs))
+            {
+                std::cout << "can't divide integers by zero!" << std::endl;
+                return *this;
+            }
+        }
+        else if (std::numeric_limits<Type>::epsilon() > static_cast<Type>(std::abs(rhs)))
+        {
+            std::cout << "warning: floating point division by zero!" << std::endl;
+        }
+        
+        *value /= static_cast<Type>(rhs);
+        return *this;
+    }
+    //========== other functions ==========
+    template <typename ArgumentType>
+    Numeric<Type>& pow(ArgumentType& rhs)
```

Comment:
Inside the templated class, you do not need to write `Numeric<Type>&` as the return type.
`Numeric&` is all you need to write.
The function parameter should be passed by const reference.
- [ ] revise